src/kernel/arch/aarch64/interrupts.zig

```diff
@@ -28,6 +28,108 @@ comptime {
+        \\.balign 0x80
+        \\ bl exceptionHandler
+        \\.balign 0x80
+        \\ bl exceptionHandler
+        \\.balign 0x80
+        \\ bl exceptionHandler
+        \\.balign 0x80
+        \\ bl exceptionHandler
+        \\.balign 0x80
+        \\ bl exceptionHandler
+        \\.balign 0x80
+        \\ bl exceptionHandler
+        \\.balign 0x80
+        \\ bl exceptionHandler
+        \\.balign 0x80
+        \\ bl exceptionHandler
+        \\.balign 0x80
+        \\ bl exceptionHandler
+        \\.balign 0x80
+        \\ bl exceptionHandler
+        \\.balign 0x80
+        \\ bl exceptionHandler
+        \\.balign 0x80
+        \\ bl exceptionHandler
+        \\.balign 0x80
+        \\ bl exceptionHandler
+        \\.balign 0x80
+        \\ bl exceptionHandler
+        \\.balign 0x80
+        \\ bl exceptionHandler
+    );
+}
+
+export fn exceptionHandler() noreturn {
+    const exception_entry_number = @truncate(u32, lr() >> 7 & 0xf);
+    var elr_elx: u64 = undefined;
+    var esr_elx: u64 = undefined;
+    var far_elx: u64 = undefined;
+    var sctlr_elx: u64 = undefined;
+    var spsr_elx: u64 = undefined;
+    var vbar_elx: u64 = undefined;
+    if (currentEl() == 3) {
+        elr_elx = mrsEl("elr_el", 3);
+        esr_elx = mrsEl("esr_el", 3);
+        far_elx = mrsEl("far_el", 3);
+        sctlr_elx = mrsEl("sctlr_el", 3);
+        spsr_elx = mrsEl("spsr_el", 3);
+        vbar_elx = mrsEl("vbar_el", 3);
+    } else if (currentEl() == 2) {
+        elr_elx = mrsEl("elr_el", 2);
+        esr_elx = mrsEl("esr_el", 2);
+        far_elx = mrsEl("far_el", 2);
+        sctlr_elx = mrsEl("sctlr_el", 2);
+        spsr_elx = mrsEl("spsr_el", 2);
+        vbar_elx = mrsEl("vbar_el", 2);
+    } else if (currentEl() == 1) {
+        elr_elx = mrsEl("elr_el", 1);
+        esr_elx = mrsEl("esr_el", 1);
+        far_elx = mrsEl("far_el", 1);
+        sctlr_elx = mrsEl("sctlr_el", 1);
+        spsr_elx = mrsEl("spsr_el", 1);
+        vbar_elx = mrsEl("vbar_el", 1);
+    }
+    log.logError("\n", .{});
+    switch (esr_elx) {
+        0x86000010 => {
+            log.logError("instruction abort\n", .{});
+            log.logError("exception level {} - no change\n", .{ currentEl()});
+            log.logError("32 bit instruction at 0x{x} accessing 0x{x}\n", .{ elr_elx, far_elx });
+        },
+        0x96000021 => {
+            log.logError("alignment fault data abort\n", .{});
+            log.logError("exception level {} - no change\n", .{ currentEl()});
+            log.logError("32 bit instruction at 0x{x} accessing 0x{x}\n", .{ elr_elx, far_elx });
+        },
+        0x96000050 => {
+            log.logError("synchronous external data abort\n", .{});
+            log.logError("exception level {} - no change\n", .{ currentEl()});
+            log.logError("32 bit instruction at 0x{x} accessing 0x{x}\n", .{ elr_elx, far_elx });
+        },
+        else => {
+            log.logError("arm exception taken\n", .{});
+        },
```

Comment:
The current system was able to determine if the interrupt was an irq, data abort, reset etc. Could that be done with this too? Perhaps using `exception_entry_number`? If you do refactor the messages for each type of exception, you could add the faulting instruction address to it as well, like it is now.

Reply:
I need to look more closely at the current system and see how it compares to the info on page 2342 of the ref manual.

Reply:
It turns out the entry for data abort maps to a synchronous reason. Data abort is one of the synchronous exceptions but there are others.

Before printing the exception details a check is made for an exception that has been observed by the programmer and if so then a specific message is printed. This helps slightly different exceptions to be noticed more easily.

Reply:
Thanks, it's good to see the specific cause of a data abort!